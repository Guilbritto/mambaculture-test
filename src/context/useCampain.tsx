'use client'

import { getCampain } from '@/app/data'
import { Campain } from '@/types/campain'
import { Paginated } from '@/types/paginated'
import React, { ReactNode, createContext, useCallback, useContext, useEffect, useState } from 'react'

type CampainContextType = {
    campains: Paginated<Campain[]>
    fetchCampains: (page?: number) => Promise<void>
}

const CampainContext = createContext<CampainContextType | undefined>(
  undefined,
)

const CampainProvider = ({ children }: { children: ReactNode }) => {
    const [campains, setCampains] = useState<Paginated<Campain[]>>({} as Paginated<Campain[]>)

    const fetchCampains = useCallback(async (page=1) => {
      const campain = await getCampain(page);
      setCampains(campain ?? {} as Paginated<Campain[]>)
  }, [])

    useEffect(() => {
      if(campains.data) return
      fetchCampains()
    }, [campains.data, fetchCampains])

    return (
        <CampainContext.Provider
            value={{ campains, fetchCampains }}
        >
        {children}
        </CampainContext.Provider>
    )
}

const useCampain = () => {
  const campainContext = useContext(CampainContext)
  if (!campainContext) {
    throw new Error('useCampain must be used within a CampainProvider')
  }
  return campainContext
}
export { CampainContext, CampainProvider, useCampain }
