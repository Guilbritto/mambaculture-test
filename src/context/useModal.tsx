'use client'

import React, { ReactNode, createContext, useContext, useState } from 'react'

type ModalContextType = {
  isOpen: boolean
  renderModal?: ReactNode
  openModal: () => void
  closeModal: () => void
  render: (renderModal: ReactNode) => void
}

const ModalContext = createContext<ModalContextType | undefined>(
  undefined,
)

const ModalProvider = ({ children }: { children: ReactNode }) => {
  const [isOpen, setIsOpen] = useState(false)
  const [renderModal, setRenderModal] = useState<
    ReactNode | undefined
  >()

  const render = (renderPageModal: ReactNode) => {
    setRenderModal(renderPageModal)
  }
  
  const openModal = () => {
    setIsOpen(true)
  }

  const closeModal = () => {
    setIsOpen(false)
    render(<></>)
  }

  return (
    <ModalContext.Provider
      value={{ isOpen, renderModal, openModal, closeModal, render }}
    >
      {children}
    </ModalContext.Provider>
  )
}
const useModal = () => {
  const modalContext = useContext(ModalContext)
  if (!modalContext) {
    throw new Error('useModal must be used within a ModalProvider')
  }
  return modalContext
}
export { ModalContext, ModalProvider, useModal }
