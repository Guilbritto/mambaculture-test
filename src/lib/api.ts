'use server'


const BASE_URL = 'http://localhost:3333';

export async function api<T>(path: string, options?: RequestInit ): Promise<T | null> {
    try{
        const response = await fetch(`${BASE_URL}/${path}`, options);
        console.log(`${BASE_URL}/${path}`);
        console.log(response);
        if (!response.ok) {
            throw new Error(response.statusText);
          }
          return await response.json() as T;
    }catch(error){
        console.error(error);
        return null;
    }
}