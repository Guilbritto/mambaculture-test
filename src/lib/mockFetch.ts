// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function mockFetch(data: any) {
    return jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => data,
      }),
    );
  }