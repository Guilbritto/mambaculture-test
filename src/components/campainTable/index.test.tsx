import React from 'react';
import { render } from '@testing-library/react';
import { CampainTable } from './index';
import { CampainProvider } from '@/context/useCampain';
import { mockFetch } from '@/lib/mockFetch';

describe('CampainTable', () => {
    beforeAll(() => {
        global.fetch = mockFetch(
            [{
                "id": 1,
                "name": "Test Campaign",
                "startDate": "2021-09-01",
                "endDate": "2021-09-30",
                "Budget": 1000,
                "userId": 1}]
        );
    });
    it('should match snapshot', () => {
        const { asFragment } = render(
            <CampainProvider>
                <CampainTable />
            </CampainProvider>
    );
        expect(asFragment()).toMatchSnapshot();
    });
});