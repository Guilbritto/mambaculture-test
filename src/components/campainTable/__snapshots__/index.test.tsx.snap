// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampainTable should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="rounded-xl border bg-card text-card-foreground shadow"
  >
    <div
      class="relative w-full overflow-auto"
    >
      <table
        class="w-full caption-bottom text-sm"
      >
        <thead
          class="[&_tr]:border-b"
        >
          <tr
            class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
          >
            <th
              class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[100px]"
            >
              Id
            </th>
            <th
              class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            >
              Nome
            </th>
            <th
              class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[200px]"
            >
              Data Inicio
            </th>
            <th
              class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[200px]"
            >
              Data Fim
            </th>
            <th
              class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[200px]"
            >
              Status
            </th>
            <th
              class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[100px]"
            >
              Ações
            </th>
          </tr>
        </thead>
        <tbody
          class="[&_tr:last-child]:border-0"
        />
      </table>
    </div>
    <nav
      aria-label="pagination"
      class="mx-auto flex w-full p-2 justify-end border-t"
      role="navigation"
    >
      <ul
        class="flex flex-row items-center gap-1"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 text-primary-foreground shadow h-9 px-4 py-2 bg-slate-700 hover:bg-slate-800"
          disabled=""
        >
          <svg
            class="lucide lucide-chevron-left"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m15 18-6-6 6-6"
            />
          </svg>
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 text-primary-foreground shadow h-9 px-4 py-2 bg-slate-700 hover:bg-slate-800"
          disabled=""
        >
          <svg
            class="lucide lucide-chevron-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </button>
      </ul>
    </nav>
  </div>
</DocumentFragment>
`;
